src/native/libs/System.Native/pal_uid.c

```diff
@@ -237,3 +238,118 @@ int32_t SystemNative_GetGroups(int32_t ngroups, uint32_t* groups)
 
     return getgroups(ngroups, groups);
 }
+
+#if !HAVE_GETGRGID_R
+// Need to call getgrgid which is not thread-safe, and protect it with a mutex
+static pthread_mutex_t s_getgrgid_lock = PTHREAD_MUTEX_INITIALIZER;
+#endif
+
+char* SystemNative_GetGroupName(uint32_t gid)
+{
+#if HAVE_GETGRGID_R
+    size_t bufferLength = 512;
+    while (1)
+    {
+        char *buffer = (char*)malloc(bufferLength);
+        if (buffer == NULL)
+        {
+            return NULL;
+        }
+
+        struct group* result;
+        struct group gr;
+        if (getgrgid_r(gid, &gr, buffer, bufferLength, &result) == 0)
+        {
+            if (result == NULL)
+            {
+                errno = ENOENT;
+                free(buffer);
+                return NULL;
+            }
+            else
+            {
+                char* name = strdup(gr.gr_name);
+                free(buffer);
+                return name;
+            }
+        }
+
+        free(buffer);
+        size_t tmpBufferLength;
+        if (errno != ERANGE || !multiply_s(bufferLength, (size_t)2, &tmpBufferLength))
+        {
+            return NULL;
+        }
+        bufferLength = tmpBufferLength;
+    }
+#else
+    // Platforms like Android API level < 24 do not have getgrgid_r available
+	pthread_mutex_lock(&s_getgrgid_lock);
+    struct group* result = getgrgid(gid);
+    if (result == NULL)
+    {
+	    pthread_mutex_unlock(&s_getgrgid_lock);
+        return NULL;
+    }
+    char* name = strdup(result->gr_name);
+    pthread_mutex_unlock(&s_getgrgid_lock);
+    return name;
+#endif
+}
+
+#if !HAVE_GETPWUID_R
+// Need to call getpwuid which is not thread-safe, and protect it with a mutex
+static pthread_mutex_t s_getpwuid_lock = PTHREAD_MUTEX_INITIALIZER;
+#endif
+
+char* SystemNative_GetUserName(uint32_t uid)
```

Comment:
Instead of creating a new System.Native export function, would it make sense to reuse `Interop.Sys.GetPwUidR`? We could refactor this code to reduce duplication:

https://github.com/dotnet/runtime/blob/6ca8c9bc0c4a5fc1082c690b6768ab3be8761b11/src/libraries/System.Private.CoreLib/src/System/Environment.Unix.cs#L27-L93

Reply:
Thank you for finding this. Your suggestion makes sense.

Reply:
I am thinking of refactoring the existing Environment code and prefer my new method that does all the passwd retrieval on native code, instead of in managed code. Maybe that could even increase perf. I'll run the benchmarks just in case, before pushing a new commit.